### PR TITLE
addpkg(main/crystal): Crystal 1.9.0

### LINKED ISSUE #3401: Package request: Crystal
It would be nice to have the crystal programming language compiler on Termux, it uses LLVM and is also self-hosted/bootstrapped but since Rust is on Termux while its also self-hosted it should also be possible, via cross-compilation and stuff.

Some hardcoded prefix paths would also need to be patched.
See also [This PR](https://github.com/crystal-lang/crystal/pull/3572) for ideas of more patches that might be needed.

Website: https://crystal-lang.org
Source: https://github.com/crystal-lang/crystal
See also for building: [Building](https://crystal-lang.org/reference/installation/from_source_repository.html), [Required Libraries](https://github.com/crystal-lang/crystal/wiki/All-required-libraries), [Cross-Compilation](https://crystal-lang.org/docs/syntax_and_semantics/cross-compilation.html)

### PR DESCRIPTION
**Pull request by HertzDevil**
This adds the [Crystal compiler](https://crystal-lang.org/) plus its Shards package manager. Only AArch64, off-device builds are tested. See also #8265

Fixes #3401

### CODE REVIEW

File: packages/crystal/build.sh [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+		$("$TERMUX_PREFIX/bin/llvm-config" --libs --system-libs --ldflags 2> /dev/null) \
+		-lstdc++ -lpcre2-8 -lm -lgc -levent -ldl
+
+	git clone --depth 1 --single-branch \
+		--branch v$SHARDS_VERSION \
+		https://github.com/crystal-lang/shards.git
+
+	cd shards
+	mkdir -p lib/molinillo
+	curl -L "$MOLINILLO_URL" | tar -xzf - -C lib/molinillo --strip-components=1
```
Comment (truboxl) [inline packages/crystal/build.sh:36]: Please use `termux_download` with checksum. You may place your downloaded file in `$TERMUX_PKG_CACHEDIR` for example.

File: packages/crystal/build.sh [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+TERMUX_PKG_DESCRIPTION="Fast and statically typed, compiled language with Ruby-like syntax"
+TERMUX_PKG_LICENSE="Apache-2.0"
+TERMUX_PKG_MAINTAINER="@HertzDevil"
+TERMUX_PKG_VERSION=1.9.0
+TERMUX_PKG_GIT_BRANCH=$TERMUX_PKG_VERSION
+TERMUX_PKG_SRCURL=git+https://github.com/crystal-lang/crystal
+TERMUX_PKG_AUTO_UPDATE=true
+_LLVM_MAJOR_VERSION=$(. $TERMUX_SCRIPTDIR/packages/libllvm/build.sh; echo $LLVM_MAJOR_VERSION)
+_LLVM_MAJOR_VERSION_NEXT=$((_LLVM_MAJOR_VERSION + 1))
+TERMUX_PKG_DEPENDS="libevent, libgc, libgmp, libiconv, libllvm (<< $_LLVM_MAJOR_VERSION_NEXT), libxml2, libyaml, openssl, pcre2, zlib"
```
Comment (truboxl) [inline packages/crystal/build.sh:11]: Are `libgmp`, `libiconv`, `libxml2`, `openssl`, `zlib` needed?
Doesn't appear to listed in DT_NEEDED for crystal and shards.

Also can you add `libc++`?

Comment (HertzDevil) [reply packages/crystal/build.sh:11]: These libraries are part of Crystal's standard library.

**Review by truboxl — APPROVED** [review]
LGTM. Lets see others got any reviews, otherwise I will merge by today.

**Comment by HertzDevil** [comment]
An addendum: `libffi` is recommended because the compiler's interpreter functionality depends on it, but it isn't exposed in the standard library API. This package doesn't enable it because the interpreter itself is an experimental feature, but people could rebuild the compiler on Termux with the source tree, enabling the interpreter, in which case `libffi` is linked against

**Pull request merged**